File: inspect_dom.py
import urllib.request
import json
import asyncio
import websockets

async def inspect():
    req = urllib.request.urlopen('http://localhost:9222/json')
    pages = json.loads(req.read().decode())
    flow_page = [p for p in pages if 'flow' in p.get('url', '')][0]
    ws_url = flow_page['webSocketDebuggerUrl']
    
    async with websockets.connect(ws_url) as ws:
        js = """
        (function() {
            var buttons = Array.from(document.querySelectorAll('button, a[role="button"], [role="button"]')).map(function(b) {
                return {
                    tag: b.tagName,
                    text: (b.innerText || b.textContent || '').trim().replace(/\\n/g, ' '),
                    aria: b.getAttribute('aria-label'),
                    visible: b.offsetParent !== null,
                    outer: b.outerHTML.substring(0, 180)
                };
            });
            var header = document.querySelector('header');
            return JSON.stringify({
                url: location.href,
                headerHtml: header ? header.outerHTML : null,
                buttons: buttons
            }, null, 2);
        })()
        """
        msg = {'id': 1, 'method': 'Runtime.evaluate', 'params': {'expression': js, 'returnByValue': True}}
        await ws.send(json.dumps(msg))
        resp = await ws.recv()
        data = json.loads(resp)
        print(data['result']['result']['value'])

asyncio.run(inspect())

File: inspect_dom.js
async function inspect() {
  const resp = await fetch('http://localhost:9222/json');
  const pages = await resp.json();
  const flowPage = pages.find(p => p.url && p.url.includes('flow'));
  const ws = new WebSocket(flowPage.webSocketDebuggerUrl);
  ws.onopen = () => {
    const js = `
      (function() {
        return JSON.stringify({
          nextData: window.__NEXT_DATA__,
          location: location.href
        }, null, 2);
      })()
    `;
    ws.send(JSON.stringify({
      id: 1,
      method: 'Runtime.evaluate',
      params: { expression: js, returnByValue: true }
    }));
  };
  ws.onmessage = (event) => {
    const data = JSON.parse(event.data);
    if (data.id === 1) {
      console.log(data.result.result.value);
      ws.close();
    }
  };
}
inspect().catch(console.error);
